test/distribution/test_nan_mixture.py

```diff
@@ -0,0 +1,344 @@
+        nll = -nan_mixture.log_prob(x)
+    nll.backward()
+
+    p_grad_true = mx.nd.where(x != x, -1 / p, 1 / (1 - p))
+
+    p_cat_zero = p_cat[:, :, :, 0]
+    p_cat_one = p_cat[:, :, :, 1]
+
+    p_cat_grad_zero_true = mx.nd.where(
+        mx.nd.logical_or(x != x, x == 1),
+        mx.nd.array(p_cat_zero).zeros_like(),
+        -1 / mx.nd.array(p_cat_zero),
+    )
+    p_cat_grad_one_true = mx.nd.where(
+        mx.nd.logical_or(x != x, x == 0),
+        mx.nd.array(p_cat_one).zeros_like(),
+        -1 / mx.nd.array(p_cat_one),
+    )
+    p_cat_grad_true = mx.nd.stack(
+        p_cat_grad_zero_true, p_cat_grad_one_true, axis=-1
+    )
+
+    assert np.allclose(p.grad.asnumpy(), p_grad_true.asnumpy(), atol=1e-1)
+    assert np.allclose(
+        p_cat.grad.asnumpy(), p_cat_grad_true.asnumpy(), atol=1e-1
+    )
+
+
+@pytest.mark.parametrize(
+    "distribution_output", [GaussianOutput(), StudentTOutput(),],
+)
+@pytest.mark.parametrize("serialize_fn", serialize_fn_list)
+def test_nanmixture_output(distribution_output, serialize_fn) -> None:
+
+    nmdo = NanMixtureOutput(distribution_output)
+
+    args_proj = nmdo.get_args_proj()
+    args_proj.initialize()
+
+    input = mx.nd.ones(shape=(3, 2))
+
+    distr_args = args_proj(input)
+
+    d = nmdo.distribution(distr_args)
+    print(d)
+    d = serialize_fn(d)
+
+    samples = d.sample(num_samples=NUM_SAMPLES)
+
+    sample = d.sample()
+
+    assert samples.shape == (NUM_SAMPLES, *sample.shape)
+
+    log_prob = d.log_prob(sample)
+
+    assert log_prob.shape == d.batch_shape
+
+
+BATCH_SIZE = 10000
+
+zeros = mx.nd.zeros((BATCH_SIZE))
+ones = mx.nd.ones((BATCH_SIZE))
+
+mu = 1.0
+sigma = 1.5
+TOL = 0.09
+
+nan_prob = 0.3
+mx.random.seed(1)
+np.random.seed(1)
+
+samples = np.random.normal(mu, scale=sigma, size=(BATCH_SIZE))
+np_samples = np.where(
+    np.random.uniform(size=(BATCH_SIZE)) > nan_prob, samples, np.nan
+)
+
+n_cat = 3
+cat_probs = np.array([0.2, 0.3, 0.5])
+cat_samples = np.random.choice(
+    list(range(n_cat)), p=cat_probs, size=BATCH_SIZE
+)
+cat_samples = np.where(
+    np.random.uniform(size=(BATCH_SIZE)) > nan_prob, cat_samples, np.nan
+)
+
+
+@pytest.mark.skip("Skip test that takes long time to run")
+def test_nanmixture_inference() -> None:
+    nmdo = NanMixtureOutput(GaussianOutput())
+
+    args_proj = nmdo.get_args_proj()
+    args_proj.initialize()
+    args_proj.hybridize()
+
+    input = mx.nd.ones((BATCH_SIZE))
+
+    trainer = mx.gluon.Trainer(
+        args_proj.collect_params(), "sgd", {"learning_rate": 0.00001}
+    )
+
+    mixture_samples = mx.nd.array(np_samples)
+
+    N = 1000
+    t = tqdm(list(range(N)))
+    for _ in t:
+        with mx.autograd.record():
+            distr_args = args_proj(input)
+            d = nmdo.distribution(distr_args)
+            loss = d.loss(mixture_samples)
+        loss.backward()
+
+        loss_value = loss.mean().asnumpy()
+        t.set_postfix({"loss": loss_value})
+        trainer.step(BATCH_SIZE)
```

Comment:
This is just an idea (and we could do this in a separate PR): we know that EM works better than SGD for the job, would it make sense to use EM in this type of tests? As long as this would involve invoking the loss (and its backward) then it should be fine. In that case we could probably hope for “stricter” assertions on the recovered parameters below

Reply:
We could do that for the Mixture distribution. However, for the NanMixture the maximisation step is not defined. Also, EM does not use the gradient. Maybe in a separate PR I could implement something like  ["A Gradient Algorithm Locally Equivalent to the EM Algorithm"](https://doi.org/10.1111/j.2517-6161.1995.tb02037.x) which uses the gradient and would, I think, work for the NanMixture.